akka-persistence-testkit/src/main/scala/akka/persistence/testkit/ProcessingPolicy.scala

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (C) 2018-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.testkit
+
+import akka.annotation.{ ApiMayChange, InternalApi }
+
+/**
+ * Policies allow to emulate behavior of the storage (failures and rejections).
+ *
+ * @tparam U type determines operations which storage can perform.
+ */
+@ApiMayChange
+trait ProcessingPolicy[U] {
```

Comment:
I was reviewing the code again and thought that maybe it is worth to add a marker trait like `trait StorageOperation` and make the type parameter to depend on that `trait ProcessingPolicy[U <: StorageOperation]`, and then extend this trait with the JounalOperation or SnapshotOperation. It seems to me that it will limit the user, and specify that Policies work with storage operations. @patriknw @raboof do you think it is worth to do?

Reply:
I have seen the question, but need to look a bit more into the policy structure before having an opinion.

Reply:
I don't think that would be needed. Seems like it would add some more complexity to these trait hierarchies without much value.

When implementing a custom ProcessingPolicy it would be like:
```
class SampleEventStoragePolicy extends EventStorage.JournalPolicies.PolicyType
```

Maybe we could make it more concrete and easier to find by providing
```
trait JournalProcessingPolicy extends ProcessingPolicy[JournalOperation]
```
and the `withPolicy` could be defined as
```
def withPolicy(policy: JournalProcessingPolicy): this.type
```

Then it's more clear that a policy for Journal PersistenceTestKit is implemented by a JournalProcessingPolicy.
```
class SampleEventStoragePolicy extends JournalProcessingPolicy
```

similar for snapshots

Reply:
I also had this idea :). Will try implement this. Last time I tried, there were some issues with it.